lookout/style/format/quality_report.py

```diff
@@ -36,19 +69,296 @@ def quality_report(input_pattern: str, bblfsh: str, language: str, n_files: int,
+        """
+        return dict(config=[deepcopy(self.config)])
+
+    def _load_tree(self, tree: dict) -> None:
+        """
+        Attach the needed data from the tree.
+
+        :param tree: asdf file tree.
+        :return: None
+        """
+        self.config = tree
+
+
+class ReportAnalyzer(FormatAnalyzer):
+    """
+    Base class for different kind of reports.
+
+    * analyze - generate report per changed files or aggregated changes.
+    * train - generate report for all files.
+
+    Child classes are required to implement 3 methods:
+    * generate_report
+    * generate_model_report (optional - by default it will return empty string)
+    * _load_analyze_config
+    """
+
+    model_type = ReportModel
+    base_config = {"aggregate": False, "model_path": None, "report_parse_failures": False}
+    defaults = {}  # has to be filled in child class.
+
+    def __init__(self, model: FormatModel, url: str, config: Mapping[str, Any]) -> None:
+        """Initialization."""
+        super().__init__(model, url, config)
+        self.config = self._load_analyze_config(self.config)
+
+    @classmethod
+    def generate_report(cls, y, y_pred, vnodes_y, target_names, rule_winners, config, model,
+                        feature_extractor) -> AnyStr:
+        """
+        Generate report. Has to be implemented in child class.
+
+        :param y: labels.
+        :param y_pred: predicted labels.
+        :param vnodes_y: virtual nodes for labels.
+        :param target_names: class names.
+        :param rule_winners: rule winner.
+        :param config: configuration.
+        :param model: pretrained model that was loaded to make report.
+        :param feature_extractor: feature extractor.
+        :return: report.
+        """
+        raise NotImplemented
+
+    @classmethod
+    def generate_model_report(cls, model, config, feature_extractor) -> AnyStr:
+        """
+        Generate report about model. Has to be implemented in child class.
+
+        :param model: model that should be described in report.
+        :param config: configuration.
+        :param feature_extractor: feature extractor.
+        :return: report.
+        """
+        return ""
+
+    @with_changed_uasts_and_contents
+    def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
+                data_request_stub: DataStub, **data) -> List[Comment]:
+        """
+        Analyze changes between revisions and make quality report per file or for all changes.
+
+        :param ptr_from: Git repository state pointer to the base revision.
+        :param ptr_to: Git repository state pointer to the head revision.
+        :param data_request_stub: Connection to the Lookout data retrieval service, not used.
+        :param data: Contains "changes" - the list of changes in the pointed state.
+        :return: List of comments.
+        """
+        self.model = FormatModel().load(self.config["model_path"])
+
+        comments = []
+
```

Comment:
I see several empty lines here and there. Better to remove them because you can fit more code on the screen.

There is some sense to insert new lines if there are really different logical parts of the function. Another option is to insert comments instead of new lines.

Reply:
Still holds